File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int helper(TreeNode* curr,int &ans) {
        if(not curr) {
            return 0;
        }
        int maxLeft=helper(curr->left,ans);
        int maxRight=helper(curr->right,ans);
        int currVal=0;
        if(curr->left) {
            maxLeft++;
        }
        if(curr->right) {
            maxRight++;
        }
        ans=max(ans,maxLeft+maxRight);
        return max(maxLeft,maxRight);
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        if(not root) {
            return 0;
        }
        int ans=0;
        helper(root,ans);
        return ans;
    }
};

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameter=0;
    public int helper(TreeNode root) {
        if(root==null) {
            return 0;
        }
        int leftHeight=helper(root.left);
        int rightHeight=helper(root.right);
        diameter=Math.max(diameter,leftHeight+rightHeight+1);
        return (Math.max(leftHeight,rightHeight)+1);
    }
    
    public int diameterOfBinaryTree(TreeNode root) {
        if(root==null) {
            return 0;
        }
        helper(root);
        return (diameter-1);
    }
}